src/ripple/conditions/impl/Fulfillment.cpp

```diff
@@ -18,144 +18,213 @@
 
-    auto const p = parsePreamble(s, ec);
-    if (ec)
-        return nullptr;
+    std::unique_ptr<Fulfillment> v;
 
-    // All fulfillments are context-specific, constructed types
-    if (!isConstructed(p) || !isContextSpecific(p))
-    {
-        ec = error::malformed_encoding;
-        return nullptr;
-    }
-
-    if (p.length > s.size())
+    der::Decoder decoder(s, der::TagMode::automatic);
+    decoder >> v >> der::eos;
+    ec = decoder.ec_;
+    if (ec)
     {
-        ec = error::buffer_underfull;
         return {};
     }
 
-    if (p.length < s.size())
-    {
-        ec = error::buffer_overfull;
-        return {};
-    }
+    return v;
+}
 
-    if (p.length > maxSerializedFulfillment)
+std::array<std::uint8_t, 32>
+Fulfillment::fingerprint(std::error_code& ec) const
+{
+    der::Encoder encoder{der::TagMode::automatic};
+    encodeFingerprint(encoder);
+    encoder << der::eos;
+
+    if (encoder.ec_)
     {
-        ec = error::large_size;
+        ec = encoder.ec_;
         return {};
     }
 
-    std::unique_ptr<Fulfillment> f;
-
-    switch (static_cast<Type>(p.tag))
-    {
-    case Type::preimageSha256:
-        f = PreimageSha256::deserialize(Slice(s.data(), p.length), ec);
-        if (ec)
-            return {};
-        s += p.length;
-        break;
-
-    case Type::prefixSha256:
-        ec = error::unsupported_type;
+    auto const& encoded = encoder.serializationBuffer(ec);
+    if (ec)
         return {};
-        break;
 
-    case Type::thresholdSha256:
-        ec = error::unsupported_type;
-        return {};
-        break;
+    sha256_hasher h;
+    h(encoded.data(), encoded.size());
+    return static_cast<sha256_hasher::result_type>(h);
+}
 
-    case Type::rsaSha256:
-        ec = error::unsupported_type;
-        return {};
-        break;
+std::bitset<5>
+Fulfillment::selfAndSubtypes() const
+{
+    std::bitset<5> result;
+    result.set(static_cast<std::size_t>(type()));
+    result |= subtypes();
+    return result;
+}
 
-    case Type::ed25519Sha256:
-        ec = error::unsupported_type;
-        return {};
+bool
+match (
+    Fulfillment const& f,
+    Condition const& c)
+{
+    // Fast check: the fulfillment's type must match the
+    // conditions's type:
+    if (f.type() != c.type)
+        return false;
 
-    default:
-        ec = error::unknown_type;
-        return {};
-    }
+    // Derive the condition from the given fulfillment
+    // and ensure that it matches the given condition.
+    std::error_code ec;
+    auto const result = c == f.condition(ec);
+    if (ec)
+        return false;
+    return result;
+}
 
-    if (!s.empty())
-    {
-        ec = error::trailing_garbage;
-        return {};
-    }
+bool
+validate (
+    Fulfillment const& f,
+    Condition const& c,
+    Slice m)
+{
+    return match (f, c) && f.validate (m);
+}
 
-    return f;
+bool
+validate (
+    Fulfillment const& f,
+    Condition const& c)
+{
+    return validate (f, c, {});
```

Comment:
Yes. Preimage validate does not depend on the message at all. Also, it's valid for a cryptocondition that depends on a message to be valid for an empty message (sign an empty message, for example).